source/extensions/common/tap/admin.cc

```diff
@@ -110,40 +109,213 @@ void AdminHandler::unregisterConfig(ExtensionConfig& config) {
+}
+
+AdminHandler::AttachedRequest::AttachedRequest(AdminHandler* admin_handler,
+                                               const envoy::admin::v3::TapRequest& tap_request,
+                                               Server::AdminStream* admin_stream)
+    : config_id_(tap_request.config_id()), config_(tap_request.tap_config()),
+      admin_stream_(admin_stream), main_thread_dispatcher_(admin_handler->main_thread_dispatcher_) {
+}
 
-    ENVOY_LOG(debug, "admin writing buffered trace to response");
-    Buffer::OwnedImpl output_buffer{output_string};
-    parent.attached_request_.value().admin_stream_->getDecoderFilterCallbacks().encodeData(
-        output_buffer, false);
+AdminHandler::AttachedRequestBuffered::AttachedRequestBuffered(
+    AdminHandler* admin_handler, const envoy::admin::v3::TapRequest& tap_request,
+    Server::AdminStream* admin_stream)
+    : AttachedRequest(admin_handler, tap_request, admin_stream) {
+  const envoy::config::tap::v3::OutputSink& sink =
+      tap_request.tap_config().output_config().sinks(0);
+
+  const uint64_t max_buffered_traces = sink.buffered_admin().max_traces();
+  const uint64_t timeout_ms = PROTOBUF_GET_MS_OR_DEFAULT(sink.buffered_admin(), timeout, 0);
+  trace_buffer_ = std::make_unique<TraceBuffer>(max_buffered_traces);
+  // Start the countdown if provided an actual timeout
+  if (timeout_ms > 0) {
+    timer_ = dispatcher().createTimer(
+        [this, admin_handler] { this->onTimeout(admin_handler->attached_request_); });
+    timer_->enableTimer(std::chrono::milliseconds(timeout_ms));
+  }
+}
+
+void AdminHandler::AttachedRequestBuffered::onTimeout(
+    const std::weak_ptr<AttachedRequest>& weak_attached_request) {
+  // Flush the buffer regardless of size
+  dispatcher().post([weak_attached_request] {
+    // Take temporary ownership - extend lifetime
+    std::shared_ptr<AttachedRequest> attached_request = weak_attached_request.lock();
+    if (!attached_request) {
+      // NOTE: Cannot do much here in response to the failed post as an HTTP response code has
+      // already been sent on completion of AdminHandler::handler. Additionally we probably don't
+      // want to take any action in this event as this case may be hit in "normal" usage, depending
+      // on when the destruction of attached_request_ occurs.
+      ENVOY_LOG(debug, "Timer Expiry after admin tap request completion");
+      return; // No attached request, flushed already.
+    }
+    TraceBuffer* trace_buffer = attached_request->traceBuffer();
+
+    // if the trace buffer has already been flushed short circuit
+    if (trace_buffer->flushed()) {
+      return;
+    }
+
+    std::vector<envoy::data::tap::v3::TraceWrapper> buffer = trace_buffer->flush();
+
+    ENVOY_LOG(debug, "Timer Expiry, admin flushing buffered traces to response");
+
+    // Serialize writes to the stream
+    for (const auto& trace : buffer) {
+      attached_request->streamMsg(trace);
+    }
+    attached_request->endStream();
```

Comment:
Related to my question above, I guess the answer is if the timer expires and there are no traces, the response will be empty. Should there be some indication that this is a timeout case so the user isn't confused?

Reply:
Yes exactly! How do you feel about adding additional explanation in the proto file as comments? 

I think putting an indicator in the response body would introduce unneeded complexity here as it would make the tests much more difficult. The issue with putting an indicator in the response is that the tests need to be able to deserialize the output as a series of traces in `PROTO_BINARY_LENGTH_DELIMITED` format. If we wanted to add something in the response body we will need to edit the API again by either adding a message or modifying the existing message used.